packages/ra-ui-materialui/src/input/AutocompleteInput.tsx

```diff
@@ -429,21 +441,29 @@ If you provided a React element for the optionText prop, you must also provide t
 
     const doesQueryMatchSelection = useCallback(
         (filter: string, eventType?: string) => {
-            let selectedItemTexts = [];
+            let selectedItemTexts;
+            let isValidChoice = false;
 
             if (multiple) {
                 selectedItemTexts = selectedChoice.map(item =>
                     getOptionLabel(item)
                 );
             } else {
                 selectedItemTexts = [getOptionLabel(selectedChoice)];
+                isValidChoice =
+                    selectedChoice &&
+                    selectedChoice[
+                        typeof optionText === 'string' ? optionText : 'name'
+                    ] === filter;
             }
 
             return eventType && eventType === 'change'
-                ? selectedItemTexts.includes(filter) && selectedChoice
+                ? selectedItemTexts.includes(filter) &&
+                      selectedChoice &&
+                      !isValidChoice
```

Comment:
I am a bit skeptical about having a `doesQueryMatchSelection` function return `false` when the choice is a valid one.

From the tests, your fix is clearly working, I'm not questioning that, but I still have the feeling we put a patch at the wrong place.
What is the purpose of this `doesQueryMatchSelection` function anyway?
To me it is very unclear what it's supposed to do, so it's difficult to tell if its implementation is correct or not.
Did you figure it out?

Reply:
Let me research a little bit, at first, removing it makes some tests fails.

Reply:
The more I think about it the more I think it's wrong :sweat_smile: 

Actually, putting the `multiple` case aside: 
- `selectedItemTexts` is equal to `[getOptionLabel(selectedChoice)]` so `selectedItemTexts.includes(filter)` is basically equivalent to `getOptionLabel(selectedChoice) === filter`
- `isValidChoice` is, after simplification, equal to `selectedChoice[optionText] === filter`, which is very very similar to `getOptionLabel(selectedChoice) === filter`

So in the end, `selectedItemTexts.includes(filter) && !isValidChoice` will probably always be `false`. We are basically expecting something and its opposite to be `true` at the same time.

Reply:
When you are right, you are right! 😄

Reply:
I simplify everything to always run the filter on input change, and call `doesQueryMatchSelection` on anything else